src/standalone/Vegetation/soil_moisture_stress.jl

```diff
@@ -0,0 +1,270 @@
+    <: AbstractSoilMoistureStressModel{FT}
+
+An implementation of a piecewise moisture stress model, taking the form
+
+βm(z) = min(1, [(θ(z)-θ_low)/(θ_high - θ_low)]^c), where θ_high,
+θ_low, and c are parameters, and θ(z) is the soil moisture at z.
+
+See  Egea et al. (2011) for details.
+
+Citation: https://doi.org/10.1016/j.agrformet.2011.05.019
+"""
+struct PiecewiseMoistureStressModel{
+    FT,
+    TMSP <: PiecewiseMoistureStressParameters{FT},
+} <: AbstractSoilMoistureStressModel{FT}
+    parameters::TMSP
+end
+
+function PiecewiseMoistureStressModel{FT}(
+    parameters::PiecewiseMoistureStressParameters{FT},
+) where {FT <: AbstractFloat}
+    return PiecewiseMoistureStressModel{eltype(parameters), typeof(parameters)}(
+        parameters,
+    )
+end
+
+ClimaLand.auxiliary_vars(model::PiecewiseMoistureStressModel) = (:βm,)
+ClimaLand.auxiliary_types(model::PiecewiseMoistureStressModel{FT}) where {FT} =
+    (FT,)
+ClimaLand.auxiliary_domain_names(::PiecewiseMoistureStressModel) = (:surface,)
+
+"""
+    compute_piecewise_moisture_stress(
+        θ_high::FT,
+        θ_low::FT,
+        c::FT,
+        θ::FT,
+    ) where {FT}
+
+This function computes at a point the soil moisture stress factor using the volumetric water content θ (m^3/m^3)
+and four parameters: `θ_high` (field capacity, m^3/m^3), `θ_low` (wilting point, m^3/m^3), `c` (curvature parameter,
+unitless). See  Egea et al. (2011).
+
+Citation: https://doi.org/10.1016/j.agrformet.2011.05.019
+"""
+function compute_piecewise_moisture_stress(
+    θ_high::FT,
+    θ_low::FT,
+    c::FT,
+    θ::FT,
+) where {FT}
+    # avoid taking e.g. sqrt of negative numbers for rational c
+    arg = max((θ - θ_low) / (θ_high - θ_low), FT(0))
+    return min(FT(1), arg^c)
+end
+
+"""
+    update_soil_moisture_stress!(
+        p,
+        Y,
+        model::PiecewiseMoistureStressModel,
+        canopy,
+    )
+
+This updates the soil moisture stress factor according to the piecewise soil moisture stress model.
+"""
+function update_soil_moisture_stress!(
+    p,
+    Y,
+    model::PiecewiseMoistureStressModel,
+    canopy,
+)
+    ground = canopy.boundary_conditions.ground
+    update_piecewise_soil_moisture_stress!(ground, p, Y, model, canopy)
+end
+
+"""
+    update_piecewise_soil_moisture_stress!(ground::PrescribedGroundConditions, p, Y, model, canopy)
+
+Updates the soil moisture stress using the piecewise model for Prescribed
+GroundConditions (p.drivers.θ prescribed).
+"""
+function update_piecewise_soil_moisture_stress!(
+    ground::PrescribedGroundConditions,
+    p,
+    Y,
+    model,
+    canopy,
+)
+    @error(
+        "You cannot use the PiecewiseSoilMoistureStress model with a prescribed soil yet."
+    )
+    #(; θ_high, θ_low, c,) = model.parameters
+    # Interpret p.drivers.θ as the root zone value.
+    #@. p.canopy.soil_moisture_stress.βm =
+    #    compute_piecewise_moisture_stress(θ_high, θ_low, c, p.drivers.θ)
```

Comment:
currently, p.drivers.psi is stored. It is another big change to convert that. let's leave that for another PR. All that means is that we cant use this new soil moisture stress parameterization with a prescribed soil model yet.